### PR TITLE
Clean up rustdoc source code

### LINKED ISSUE #70498: Clean up src/librustdoc/markdown.rs render function
It should return a `Result`/`Option` and not just an integer, it makes the code a bit confusing.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #70498.

r? @kinnison 

cc @rust-lang/rustdoc

### CODE REVIEW

File: src/librustdoc/markdown.rs [hunk 1]
```diff
@@ -55,26 +54,22 @@ pub fn render(
         css.push_str(&s)
     }
 
-    let input_str = match load_string(&input, diag) {
+    let input_str = match load_string(input, diag) {
         Ok(s) => s,
-        Err(LoadStringError::ReadFail) => return 1,
-        Err(LoadStringError::BadUtf8) => return 2,
+        Err(LoadStringError::ReadFail) => return Err("Read error: read failed".to_owned()),
+        Err(LoadStringError::BadUtf8) => return Err("Read error: invalid UTF8".to_owned()),
```
Comment (ollie27) [inline src/librustdoc/markdown.rs:60]: `load_string` already prints error messages so this results in duplicated messages. We can probably just use `std::fs::read_to_string` here.

File: src/librustdoc/markdown.rs [hunk 2]
```diff
@@ -35,14 +35,13 @@ fn extract_leading_metadata(s: &str) -> (Vec<&str>, &str) {
 /// Render `input` (e.g., "foo.md") into an HTML file in `output`
 /// (e.g., output = "bar" => "bar/foo.html").
 pub fn render(
-    input: PathBuf,
+    input: &PathBuf,
```
Comment (kinnison) [inline src/librustdoc/markdown.rs:38]: Should this not be a `&Path` or even an `AsRef<Path>` to be more idiomatic?

File: src/librustdoc/markdown.rs [hunk 3]
```diff
@@ -55,26 +54,22 @@ pub fn render(
...
     let playground = playground_url.map(|url| markdown::Playground { crate_name: None, url });
 
     let mut out = match File::create(&output) {
-        Err(e) => {
-            diag.struct_err(&format!("{}: {}", output.display(), e)).emit();
-            return 4;
-        }
+        Err(e) => return Err(format!("{}: {}", output.display(), e)),
         Ok(f) => f,
     };
```
Comment (kinnison) [inline src/librustdoc/markdown.rs:68]: Maybe something like:

```rust
let mut out = File::create(&output).map_err(|e| format!("{}: {}", output.display(), e))?;
```

**Review by kinnison — COMMENTED** [review]
One concern I have is that we've now changed the CLI ABI-incompatibly in that we no longer have distinguished exit codes for the various error conditions.  Is this going to be a problem for anyone?

File: src/librustdoc/markdown.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/librustdoc/markdown.rs:38]: Absolutely!

File: src/librustdoc/markdown.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply src/librustdoc/markdown.rs:68]: :+1:

**Comment by GuillaumeGomez** [comment]
> One concern I have is that we've now changed the CLI ABI-incompatibly in that we no longer have distinguished exit codes for the various error conditions. Is this going to be a problem for anyone?

No one uses rustdoc CLI ABI except rustdoc itself. So if there is a problem, the current tests will catch it.

File: src/librustdoc/markdown.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustdoc/markdown.rs:60]: Good point, I'll check how to prevent it.

**Comment by GuillaumeGomez** [comment]
Updated!

**Review by kinnison — APPROVED** [review]
Looks reasonable to me.  If @ollie27 agrees then r+

File: src/librustdoc/docfs.rs [hunk 4]
```diff
@@ -12,6 +12,7 @@
 use std::fs;
 use std::io;
 use std::path::Path;
+use std::string::ToString;
```
Comment (ollie27) [inline src/librustdoc/docfs.rs:15]: `ToString` is in the prelude so doesn't need to be imported.

**Comment by ollie27** [comment]
> No one uses rustdoc CLI ABI except rustdoc itself.

I don't know what that means but this PR does change the exit codes of rustdoc which people could be relying on. I hope no one is because they don't seem to be documented so I think this is fine to change.

> So if there is a problem, the current tests will catch it.

Do we even have any tests for these error cases?

**Comment by Mark-Simulacrum** [comment]
Please be more cautious about breakage like this :)

However in this case I agree that we should be free to change exit codes (so long as all errors still have nonzero exit status). However, we should be sure to mark as relnotes so that this can get recorded in compat notes.

@rustbot label relnotes

**Comment by ollie27** [comment]
`markdown::test` also returns exit codes that aren't 0 or 1 so it will be worth changing that as well to keep things consistent: https://github.com/rust-lang/rust/blob/7ebd87a7a1e0e21767422e115c9455ef6e6d4bee/src/librustdoc/markdown.rs#L138

**Comment by GuillaumeGomez** [comment]
I'll make the change then!

**Comment by GuillaumeGomez** [comment]
Updated!

File: src/librustdoc/docfs.rs (hunk 4 above)
Comment (ollie27) [reply src/librustdoc/docfs.rs:15]: @GuillaumeGomez did you miss this comment?

**Pull request merged**